presentation-compiler/src/main/dotty/tools/pc/printer/ShortenedTypePrinter.scala

```diff
@@ -0,0 +1,546 @@
+        case _ =>
+          Seq.empty
+
+    def label(paramss: List[List[Symbol]]) = {
+      var index = 0
+      paramss.flatMap { params =>
+        val labels = params.flatMap { param =>
+          // Don't show implicit evidence params
+          // e.g.
+          // from [A: Ordering](a: A, b: A)(implicit evidence$1: Ordering[A])
+          // to   [A: Ordering](a: A, b: A): A
+          val lab =
+            if implicitEvidenceParams.contains(param) then Nil
+            else
+              paramLabel(
+                param,
+                implicitEvidencesByTypeParam,
+                index,
+                if !onlyMethodParams then paramsDocs else Seq.empty,
+                nameToInfo
+              ) :: Nil
+          index += 1
+          lab
+
+        }
+        // Remove empty params
+        if labels.isEmpty then Nil
+        else labels.iterator :: Nil
+      }
+    }.iterator
+    val paramLabelss = label(methodParams)
+    val extLabelss = label(extParams)
+
+    val returnType = tpe(gtpe.finalResultType)
+    def extensionSignatureString =
+      val extensionSignature = paramssString(extLabelss, extParams)
+      if extParams.nonEmpty then
+        extensionSignature.mkString("extension ", "", " ")
+      else ""
+    val paramssSignature = paramssString(paramLabelss, methodParams)
+      .mkString("", "", s": ${returnType}")
+
+    val flags = (gsym.flags & methodFlags)
+    val flagsSeq =
+      if !flags.isEmpty then
+        val privateWithin =
+          if gsym.privateWithin != NoSymbol then gsym.privateWithin.name.show
+          else ""
+        flags.flagStrings(privateWithin)
+      else Nil
+    val mods = (additionalMods ++ flagsSeq).distinct match
+      case Nil => ""
+      case xs => xs.mkString("", " ", " ")
+
+    if onlyMethodParams then paramssSignature
+    else
+      // For Scala2 compatibility, show "this" instead of <init> for constructor
+      val name = if gsym.isConstructor then StdNames.nme.this_ else gsym.name
+      extensionSignatureString +
+        s"${mods}def $name" +
+        paramssSignature
+  end defaultMethodSignature
+
+  def defaultValueSignature(
+      gsym: Symbol,
+      gtpe: Type,
+      additionalMods: List[String] = Nil
+  ): String =
+    val flags = (gsym.flags & methodFlags)
+    val flagString =
+      if !flags.isEmpty then
+        val privateWithin =
+          if gsym.privateWithin != NoSymbol then gsym.privateWithin.name.show
+          else ""
+        flags.flagStrings(privateWithin).mkString(" ") + " "
+      else ""
+    val mods =
+      if additionalMods.isEmpty then flagString
+      else additionalMods.mkString(" ") + " " + flagString
+    val prefix = if gsym.is(Mutable) then "var" else "val"
+    s"${mods}$prefix ${gsym.name.show}: ${tpe(gtpe)}"
+  end defaultValueSignature
+
+  /*
+   * Check if a method is an extension method and in that case separate the parameters
+   * into 2 groups to make it possible to print extensions properly.
+   */
```

Comment:
I'm pretty scaladoc also has logic for pretty-printing extension methods, so we should also try to combine these things at some point in the future.

Reply:
I’ve created https://github.com/lampepfl/dotty/issues/18145 to track that.